### PR TITLE
[InstCombine] fold `cond ? x : -x == 0` into `x == 0`

### LINKED ISSUE #85250: missed FP fold, (x >= 0) ? x : -x == 0 => x == 0
https://alive2.llvm.org/ce/z/NFEKjp
https://godbolt.org/z/7Ef9Kf4s3
```llvm
define i1 @src(float %0) {
entry:
%1 = fcmp oge float %0, 0.000000e+00
%2 = fneg float %0
%3 = select i1 %1, float %0, float %2
%4 = fcmp oeq float %3, 0.000000e+00
ret i1 %4
}

define i1 @tgt(float %0) {
entry:
%oeq = fcmp oeq float %0, 0.000000e+00
ret i1 %oeq
}
```
@jayfoad @regehr

### PR DESCRIPTION
**Pull request by SahilPatidar**
Resolve #85250 
Alive2: https://alive2.llvm.org/ce/z/7DMRCy

### CODE REVIEW

**Comment by SahilPatidar** [comment]
@dtcxzyw, please check this out before making it ready for review.

**Comment by SahilPatidar** [comment]
@jayfoad, @arsenm

File: llvm/test/Transforms/InstCombine/fcmp.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/fcmp.ll:1378]: Don't use fast flag.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7980]: We should propagate fast-math flags.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7976]: Alive2: https://alive2.llvm.org/ce/z/7DMRCy

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7978]: Please handle the commuted version: select Cond, -X, X.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7979]:
```suggestion
          match(LHSI, m_Select(m_Value(), m_Value(X), m_FNeg(m_Specific(X)))) && match(RHSC, m_AnyZeroFP()))
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7977]:
```suggestion
      if (FCmpInst::isEquality(Pred) &&
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (jayfoad) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7979]:
```suggestion
          match(LHSI, m_Select(m_Value(), m_Value(X), m_FNeg(m_Deferred(X)))) &&
          match(RHSC, m_AnyZeroFP()))
```

Comment (jayfoad) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7979]: Also you should check the RHS constant before the LHS expression, since it's cheaper to check.

File: llvm/test/Transforms/InstCombine/fcmp.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fcmp.ll:1393]: Tests with -0s

File: llvm/test/Transforms/InstCombine/fcmp.ll
Comment (elhewaty) [inline llvm/test/Transforms/InstCombine/fcmp.ll:1]: Can you please split the patch into two separate files, so it's easier to see the diff?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (elhewaty) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7979]: wouldn't it be better if it's `m_OneUse(m_FNeg(m_Deferred(X)))`?

Comment (jayfoad) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7979]: No - the transform is still beneficial even if we can't remove the fneg instruction.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (SahilPatidar) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7976]: could I use this `computeKnownFPClass(Op1).isKnownNever(FPClassTest::fcNan | FPClassTest::fcInf | FPClassTest::fcNormal | FPClassTest::fcSubnormal)`?

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7976]: No. All you need is `m_AnyZero`.

Comment (SahilPatidar) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7976]: I'm concerned this (half nofpclass(inf nan norm sub) %y) might cause test failures.

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7976]: > I'm concerned this (half nofpclass(inf nan norm sub) %y) might cause test failures.

It is for alive2 proof, but not for tests.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (jayfoad) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7976]:
```suggestion
      // fcmp eq (cond ? x : -x), 0 --> fcmp eq x, 0
      if (FCmpInst::isEquality(Pred) && match(RHSC, m_AnyZeroFP()) &&
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7980]: Could you please add some FMF tests?

File: llvm/test/Transforms/InstCombine/fcmp.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/fcmp.ll:1420]: Please add a test with +/-0 as well.
https://alive2.llvm.org/ce/z/qhk2Ry
```
define <2 x i1> @src(<2 x i1> %cond, <2 x float> %x) {
  %fneg = fneg <2 x float> %x
  %sel = select <2 x i1> %cond, <2 x float> %x, <2 x float> %fneg
  %icmp = fcmp oeq <2 x float> %sel, <float 0.0, float -0.0>
  ret <2 x i1> %icmp
}
```

File: llvm/test/Transforms/InstCombine/fcmp.ll
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fcmp.ll:1441]: And a test with denormal-fp-math=dynamic,dynamic

Comment (SahilPatidar) [reply llvm/test/Transforms/InstCombine/fcmp.ll:1518]: I'm not familiar with this what it is flag in opt.

Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fcmp.ll:1518]: It's an attribute, see the LangRef

Comment (SahilPatidar) [reply llvm/test/Transforms/InstCombine/fcmp.ll:1518]: I used -denormal-fp-math=dynamic in comment script, but I didn't see any difference in these tests.

Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fcmp.ll:1518]: I don't expect to see a difference, but alive should catch violations if it's wrong

Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fcmp.ll:1518]: I also think you should just directly add the attribute to a reference test or 2, not use the opt flag

Comment (SahilPatidar) [reply llvm/test/Transforms/InstCombine/fcmp.ll:1518]: I apologize for the delay. I tried adding the flag "denormal-fp-math=dynamic" after function signature but I didn't see any difference.

**Comment by SahilPatidar** [comment]
@dtcxzyw

File: llvm/test/Transforms/InstCombine/fcmp.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/fcmp.ll:1432]: You only need one test for vec/fmf separately.

File: llvm/test/Transforms/InstCombine/fcmp.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/fcmp.ll:1494]: Please use individual flags (e.g., nnan/ninf) instead of `fast`.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (jayfoad) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8025]: This will drop FMF and the name of the original fcmp inst. How about:
```suggestion
        return replaceOperand(I, 0, X);
```

File: llvm/test/Transforms/InstCombine/fcmp.ll
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fcmp.ll:1518]: Correct. You should not see a difference. you should still have the test showing that it does not make a difference. Also, it should be "dynamic,dynamic" for the value

**Review by arsenm — APPROVED** [review]

**Comment by arsenm** [comment]
Needs a resolve

**Comment by SahilPatidar** [comment]
> Needs a resolve

Could you please tell me what specific issue needs to be resolved?

**Comment by nikic** [comment]
@SahilPatidar I think @arsenm meant rebase, not resolve :) There is a merge conflict.

**Pull request merged**
